lib/fs/basicfs_case.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+)
+
+type ErrCase struct {
+	given, real string
+}
+
+func (e *ErrCase) Error() string {
+	return fmt.Sprintf(`given name "%v" differs from name in filesystem "%v"`, e.given, e.real)
+}
+
+func IsErrCase(err error) bool {
+	e := &ErrCase{}
+	return errors.As(err, &e)
+}
+
+// caseBasicFilesystem is a BasicFilesystem with additional checks to make a
+// potentially case insensitive underlying FS behave like it's case-sensitive.
+type caseBasicFilesystem struct {
```

Comment:
Nothing makes this bound to basicFilesystem (except for the windows specialization), so this should just be `caseSensitiveFilesystem` encapsulating a general underlying fs.Filesystem.


The newRealCaser is then passed underlying fs.Filesystem and can have a special case:
```
if underlying.Type() == Basic && runtime.GOOS == "windows" {
   return windowsBasedRealCaser
} else {
   return generalRealCaserThatWorksOnAnyFilesysmte
}
```

Reply:
And then you add a `makeCaseSensitive bool` argument to `NewFilesystem`?

Reply:
So it depends on how we spin this.

Currently we default to wrapping everything, with the option to switch it off.
We could make the logic for wrapping happen in NewFilesystem based on some runtime.GOOS logic.
We could make it an option on the folder config I guess, which we then push down to NewFilesystem
Or we could have a magic type.

The thing I don't like about the magic type is that it's not an implementation of it's own, it's just a wrapper around an existing implementation. Now if we ever add let's say s3 as an implementation, you'll have to have caseS3 etc as different types that act just as wrappers and don't provide any functionality themselves.

We have walkFs and mtimeFs which perform a simillar function, but there are not types you can instantiated.

Sure, they are somewhat not optional, hence I am questioning whether this should be, and if yes, under what circumstances.

Reply:
I am now spinning it similar to mtimefs, i.e. casefs is now filesystem agnostic and you get it by calling `fs.NewCaseFilesystem(fs Filesystem)`. That happens in `FolderConfiguration.Filesystem()`, depending on `FolderConfiguration.CaseSensitiveFS` (which is false by default, meaning case resolving happens). I expect this to be a checkbox in the edit folder modal in the advanced tab.